padertorch/ops/stft_module.py

```diff
@@ -0,0 +1,270 @@
+from paderbox.transform.module_stft import sample_index_to_stft_frame_index
+from paderbox.transform.module_stft import _biorthogonal_window_fastest
+from scipy.signal.windows import blackman
+import typing
+from math import ceil
+
+from torch.nn import functional as F
+
+
+def get_stft_kernel(size, window):
+    length = len(window)
+    real = np.array([
+        [np.cos(-1 * n * 2 * np.pi / size * k) * window[k]
+         for k in range(length)] for n in range(size // 2 + 1)
+    ])
+    imag = np.array([
+        [np.sin(-1 * n * 2 * np.pi / size * k) * window[k]
+         for k in range(length)] for n in range(size // 2 + 1)
+    ])
+    kernel = np.concatenate([real, imag], axis=0)
+    # print(kernel.shape)
+    return torch.from_numpy(kernel).unsqueeze(dim=1)
+
+
+def get_istft_kernel(size, shift, window):
+    window = _biorthogonal_window_fastest(window, shift) / size
+    length = len(window)
+    kernel_numpy = np.array([
+        [np.cos(1 * f * 2 * np.pi / size * n) * window[n]
+         for n in range(length)] for f in range(size)
+    ])
+    kernel_real = torch.from_numpy(kernel_numpy)
+    kernel_real = torch.unsqueeze(kernel_real, dim=1)
+
+    kernel_numpy = np.array([
+        [np.sin(-1 * f * 2 * np.pi / size * n) * window[n]
+         for n in range(length)] for f in range(size)
+    ])
+    kernel_imag = torch.from_numpy(kernel_numpy)
+    kernel_imag = torch.unsqueeze(kernel_imag, dim=1)
+    return kernel_real, kernel_imag
+
+
+class STFT:
+    def __init__(
+            self,
+            size: int = 1024,
+            shift: int = 256,
+            *,
+            window: [str, typing.Callable] = blackman,
+            window_length: int = None,
+            fading: typing.Optional[typing.Union[bool, str]] = 'full',
+            pad: bool = True,
+            symmetric_window: bool = False,
+            complex_representation: str = 'complex'
+    ):
+        """
+        This is a torch stft implementation which mirrors the behavior of
+        the numpy implementation in paderbox.transform.stft.
+        For additional informations about the parameters please refer to
+        the numpy implementation.
+        Args:
+
+            size: Scalar FFT-size.
+            shift: Scalar FFT-shift. Typically shift is a fraction of size.
+            window: Window function handle.
+            window_length:
+            fading:
+            pad:
+            symmetric_window:
+            complex_representation: str, defines how to handle the real and
+                imaginary part of the complex stft signal:
+                                either complex, concat or stacked
+                                complex is not supported at the moment
+        """
+        possible_out_types = ['complex', 'concat', 'stacked']
+        assert complex_representation in possible_out_types, (
+            f'Please choose one of the predefined output_types'
+            f' {possible_out_types}'
```

Comment:
Can you add `complex_representation` in the assertion msg?

Reply:
done